### ISSUE TITLE
Access memory violation / Game Crashes

### ISSUE THREAD

**Issue by Grabowski95** [niemand-sec/LuaHook#1]
Hello,

I find your lua hook very useful and I wanted to test it on modded vermintide 2 realm with EAC disabled.
I receive random crashes and spent several hours straight trying to understand whats going on.
It starts correctly with lua **state = 0 and new lua state = 0**  (game offset correct)
Clicking Home (changed hotkey from F9)...
Getting 0 0 again, worked, fire mage boosted levels.
2nd time, same
3rd time, lua state=0, new lua state=4 and getting memory access violation and address like 0x0000000000010 for instance

I have few experience with lua. I tried hooking lua_settop and call **val.lua_settop_p(lua_State_ptr, 0);**  to clear stack but without a luck.
Also, checked creating new lua thread whens its value=4.
Sometimes i sucessfuly execute lua like 5 times in a row and crash and sometimes crash after 1st or 2nd.

May you help me? 
Can we catch these exceptions (despite lua is written in c) and take action or continue loop?

Cheers,
Filip

**Comment by niemand-sec** [comment]
Hi Filip, I'm glad you have found this project useful :) Several months have passed from the date I released it, many things may have changed on the game but the hook should still be working. 

The first thing that comes to my mind are race conditions. Remember that the LuaState is an object that is being used every time a lua function is called (many times per second) and it is being constantly modified. I'm not sure what are you executing on each try, but probably if the LuaState is not on the correct "state" for that call, that could probably cause the crash.

A second thing you could try it is to create a new state based on the original one before performing each call, not sure if this will work or it is a good solution. It has been a while since the last time I played with lua.

**Comment by Grabowski95** [comment]
I managed to make it working again 👍 
I turned out that fking game instantly deletes my lua threads before i can make pcall
Also, I played a bit with creating newstance and loadlibs() and it also works but cant execute custom scripts as game libraries are not loaded.
Once again, thanks. This ordered my knowledge with hooks and a bit with ida.

**Comment by niemand-sec** [comment]
Feel free to drop some snipped of code here if you want. Someone may find it useful :)

How did you do to keep your threads alive?

**Comment by Grabowski95** [comment]
Okay so the drill is as follows:
Game indeed was deleting my lua threads so I was making a new one every time in getasynckeystate loop but with a appropriate lua scripts.
I used vermintide 2 developer console to get some debug info so I knew threads were deleted after execution.
But it wasnt enough - I mean it survived a bit longer but the issue was also race condition for lua state. 
I tried with creating whole new state and there was no problem with execution except that had no access to game functions.
Finally I hooked some lua function to prevent execution overlapping (so 1 at a time) and instead i use they origin state - works like a charm.
